src/features/device-state/routes/state.ts

```diff
@@ -301,28 +367,83 @@ export const stateV2: RequestHandler = async (req, res) => {
 	const { apiKey } = req;
 	events.emit('get-state', uuid, { apiKey });
 
+	try {
+		const device = await getDevice(req, uuid);
+		const config = getConfig(device) ?? {};
+
+		const appsForState: Dictionary<AppV2> = {};
+
+		const userApp = getUserAppForState(device, config);
+		if (userApp) {
+			const userAppV2 = {
+				commit: userApp.commit,
+				name: userApp.name,
+				releaseId: userApp.releaseId,
+				networks: userApp.networks,
+				services: userApp.services,
+				volumes: userApp.volumes,
+			} as AppV2;
+			appsForState[userApp.appId] = userAppV2;
+		}
+
+		const local: LocalStateV2 = {
+			name: device.device_name ?? '',
+			config,
+			apps: appsForState,
+		};
+
+		const dependent = await getDependent(device);
+
+		res.json({
+			local,
+			dependent,
+		} as ResponseV2);
+	} catch (err) {
+		if (handleHttpErrors(req, res, err)) {
+			return;
+		}
+		captureException(err, 'Error getting device state', { req });
+		res.sendStatus(500);
+	}
+};
+
+export const stateV3: RequestHandler = async (req, res) => {
```

Comment:
There's a lot of repetition between the two handlers. I think it's best to extract the surrounding common bits into a single function and pass most of the `try` block as a callback. Something like this should work:

```ts
const makeStateHandler = (cb: (uuid, apiKey) => Promise<AnyObject>) => async (req, res) => {
	...

	try {
		const result = await cb(uuid, apikey);
		res.json(...)
	} catch {
		...
	}
}

const stateV2: RequestHandler = makeStateHandler(() => {
	...
});
```

This will still preserve quite a bit of repetition but it's the first obvious step towards DRY'ing this up a bit.

All that said, I don't feel strongly about doing this now -- feel free to merge but please make sure to return and improve this soon after.

Reply:
The previous refactoring (previous to the current PR's) went a long way toward avoiding WET code.  They are somewhat similar, but I felt that trying to coalesce them any further would lead to code that is hard to read.  We can look at this again once this PR is done.

Reply:
resolving since Akis is out.